packages/flutter/lib/src/foundation/timeline.dart

```diff
@@ -0,0 +1,395 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:developer';
+import 'dart:typed_data';
+
+import 'package:meta/meta.dart';
+
+import '_timeline_io.dart'
+  if (dart.library.js_util) '_timeline_web.dart' as impl;
+
+/// Measures how long blocks of code take to run.
+///
+/// This class can be used as a drop-in replacement for [Timeline] as it
+/// provides methods compatible with [Timeline] signature-wise, and it has
+/// minimal overhead.
+///
+/// Provides [reset] and [collect] methods that make it convenient to use in
+/// frame-oriented environment where collected metrics can be attributed to a
+/// frame, then aggregated into frame statistics, e.g. frame averages.
+///
+/// Forwards measurements to [Timeline] so they appear in Flutter DevTools.
+abstract final class FlutterTimeline {
+  static _BlockBuffer _buffer = _BlockBuffer();
+
+  /// Whether block timings are collected and can be retrieved using the
+  /// [collect] method.
+  static bool get collectionEnabled => _collectionEnabled;
+
+  /// Enables metric collection.
+  ///
+  /// Metric collection can only be enabled in non-release modes. It is most
+  /// useful in the profile mode where application performance is representative
+  /// of a deployed application.
+  ///
+  /// When disabled, resets collected data by calling [reset].
+  ///
+  /// Throws a [StateError] if invoked in release mode.
+  static set collectionEnabled(bool value) {
+    if (value == _collectionEnabled) {
+      return;
+    }
+    _collectionEnabled = value;
+    reset();
+  }
+
+  static bool _collectionEnabled = false;
+
+  /// Start a synchronous operation labeled [name].
+  ///
+  /// Optionally takes a map of [arguments]. This slice may also optionally be
+  /// associated with a [Flow] event. This operation must be finished by calling
+  /// [finishSync] before returning to the event queue.
+  static void startSync(String name, { Map<String, Object?>? arguments, Flow? flow }) {
+    Timeline.startSync(name, arguments: arguments, flow: flow);
+    if (_collectionEnabled) {
```

Comment:
all these should probably also be guarded by !kReleaseMode to ensure that they are compiled out in release mode to lower the overhead even further.

Reply:
Done.